### PR TITLE
@W-14662364: [Android] Add additional Native Login Feature: User Registration

### PR DESCRIPTION
**Pull request by JohnsonEricAtSalesforce**
🎸 _Review For Review!_ 🥁

Now we have headless registration via the Salesforce Identity API as well.  This will look very familiar to reviewers of other recent headless identity flow pull requests.

### CODE REVIEW

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt [hunk 1]
```diff
@@ -35,8 +35,10 @@ import android.util.Base64.NO_PADDING
 import android.util.Base64.NO_WRAP
 import android.util.Base64.URL_SAFE
 import android.util.Base64.encodeToString
+import android.util.Patterns.EMAIL_ADDRESS
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:38]: We'll see this later for the new email input validation routine.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt [hunk 2]
```diff
@@ -266,6 +271,146 @@ internal class NativeLoginManager(
...
+        firstName: String,
+        lastName: String,
+        username: String,
+        newPassword: String,
+        reCaptchaToken: String,
+        otpVerificationMethod: OtpVerificationMethod
+    ): StartRegistrationResult {
+        // Validate parameters.
+        val trimmedEmail = email.trim()
+        if (!EMAIL_ADDRESS.matcher(trimmedEmail).matches()) {
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:287]: For email, it seemed easy enough to use Android's provided email pattern.  Any objections?

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt [hunk 3]
```diff
@@ -266,6 +271,146 @@ internal class NativeLoginManager(
         return requestBodyString.toRequestBody(mediaType)
     }
 
+    // region Salesforce Identity API Headless Registration Flow
+
+    override suspend fun startRegistration(
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:276]: Everything here in start registration should seem very similar to all the previous native authentication methods.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt [hunk 4]
```diff
@@ -266,6 +271,146 @@ internal class NativeLoginManager(
...
+                StartRegistrationResult(UnknownError)
+            }
+        }
+    }
+
+    override suspend fun completeRegistration(
+        otp: String,
+        requestIdentifier: String,
+        otpVerificationMethod: OtpVerificationMethod
+    ) = submitAuthorizationRequest(
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:356]: Complete registration is identical to complete password-less authorization, so here's a common method extracted to fulfill both.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt [hunk 5]
```diff
@@ -266,6 +271,146 @@ internal class NativeLoginManager(
...
+        /**
+         * A data class for the Salesforce Identity API start registration
+         * request body's user info parameter.
+         * @param email A valid, user-entered email address
+         * @param username A valid Salesforce username or email
+         * @param password The user-entered new password
+         * @param firstName The user-entered first name
+         * @param lastName The user-entered last name
+         */
+        data class UserData(
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:396]: I subtle difference here from previous methods is that we have an object type as one of the fields, so here's that type.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt [hunk 6]
```diff
@@ -443,60 +588,12 @@ internal class NativeLoginManager(
         otp: String,
         otpIdentifier: String,
         otpVerificationMethod: OtpVerificationMethod
-    ): NativeLoginResult {
-        // Validate parameters.
-        val trimmedOtp = otp.trim()
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:448]: All of this code is now common in `submitAuthorizationRequest`.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt [hunk 7]
```diff
@@ -673,6 +770,78 @@ internal class NativeLoginManager(
...
+     * Identity API
+     * @param identifier A OTP or request identifier previously issued by the
+     * Salesforce Identity API to match the provided OTP
+     * @param otpVerificationMethod The OTP verification method used to obtain
+     * the identifier from the Salesforce Identity API
+     * @return A native login result indicating success or one of several
+     * possible failures, including both in-app and Salesforce Identity API
+     * results
+     */
+    private suspend fun submitAuthorizationRequest(
```
Comment (JohnsonEricAtSalesforce) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:788]: This code should be 99.9% identical to what's on `dev` with the exception of the `authRequestType` parameter.

**Review by JohnsonEricAtSalesforce — COMMENTED** [review]
The new code for registration is very similar to existing patterns from login, password-less login and forgot password, so there's just a few quick walkthrough comments here 😎

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt [hunk 8]
```diff
@@ -266,6 +271,146 @@ internal class NativeLoginManager(
...
+        }.onFailure { e ->
+            SalesforceSDKLogger.e(TAG, "Cannot JSON encode start registration request body due to an encoding error with message '${e.message}'.", e)
+            return StartRegistrationResult(UnknownError)
+        }.getOrNull()
+
+        // Create the start registration request.
+        val startRegistrationRequest = RestRequest(
+            POST,
+            LOGIN,
+            "$loginUrl$OAUTH_ENDPOINT_HEADLESS_INIT_REGISTRATION",
```
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt:321]: NIT:  I know I have said this before, but I truly think this is unreadable.  I don't think this is what the designers or doc writers had in mind when said to prefer string templates over concatenation.  Ever single example they give has string literals and spaces around the variables used.  You wouldn't pass `"$loginUrl"` as a parameter, right?

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/interfaces/NativeLoginManager.kt [hunk 9]
```diff
@@ -72,6 +82,70 @@ interface NativeLoginManager {
      */
     fun getFallbackWebAuthenticationIntent(): Intent
 
+    // region Salesforce Identity API Headless Registration Flow
+
+    /**
+     * Submits a request to start a user registration to the Salesforce Identity
+     * API headless registration flow.  This fulfills step four of the headless
+     * registration flow.
```
Comment (brandonpage) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/interfaces/NativeLoginManager.kt:90]: You probably want to link the documentation this is referencing (like you did with password reset) for this to make sense to the reader.   Same with `completeRegistration`.

**Review by brandonpage — APPROVED** [review]

**Review by wmathurin — APPROVED** [review]

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/interfaces/NativeLoginManager.kt (hunk 9 above)
Comment (JohnsonEricAtSalesforce) [reply libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/interfaces/NativeLoginManager.kt:90]: 🦅 I award @brandonpage the "Eagle Eyes" award for noticing that!

**Pull request merged**